lib/m2ts/caption-stream.js

```diff
@@ -319,77 +494,162 @@ var Cea608Stream = function() {
+          this.addFormatting(packet.pts, 'u');
+        }
+
+      // Detect offset control codes, but don't affect output
+      } else if (char0 === this.offset_ && (char1 >= 0x21 && char1 <= 0x23)) {
+        this.column_ += (char1 & 0x03);
+        return; // No text will be added to the caption, so just return now
+
+      // Detect PACs (Preamble Address Codes)
+      } else if (char0 >= this.base_ && char0 < (this.base_ + 8)) {
+
+        // There's no logic for PAC -> row mapping, so we have to just
+        // iterate through an array and compare :(
+        ROWS.forEach(function(baseCode, index) {
+          if ((data & 0x1f20) === baseCode) {
+            if (index !== this.row_) {
+              // formatting is only persistent for current row
+              this.clearFormatting(packet.pts);
+              this.row_ = index;
+            }
+
+            if ((data & 0x10) === 0x10) {
+              // we've got an indent level
+              this.column_ = ((data & 0xe) >> 1) * 4;
+            }
+
+            return; // we've found the row so break out of this function
+
+          }
+        }, this);
+
+        // All PACs can apply underline, so detect and apply
+        if ((char1 & 0x1) && (this.formatting_.indexOf('u') === -1)) {
+            this.addFormatting(packet.pts, 'u');
+        }
+
+        if ((char1 >= 0x40 && char1 < 0x50) || (char1 >= 0x60 && char1 < 0x70)) {
+          // it's a color code, though we only support white, which
+          // can be either normal or italicized.
+          if ((char1 & 0xe) === 0xe) {
+            this.addFormatting(packet.pts, 'i');
+          }
+        }
+
+        // No text will be added to the caption, so just return now
+        return;
 
-      // Look for special character sets
-      if ((char0 === 0x11 || char0 === 0x19) &&
-          (char1 >= 0x30 && char1 <= 0x3F)) {
-        // Put in eigth note and space
-        char0 = 0x266A;
-        char1 = '';
-      }
+      // We have a normal character in char0, and possibly one in char1
+      } else if (char0 >= 0x20 && char0 <= 0x7f) {
+        if (char1 === 0x00) {
+          char1 = null;
+        }
+        text += getCharFromCode(char0);
+        text += getCharFromCode(char1);
+        this.column_ += text.length;
 
-      // ignore unsupported control codes
-      if ((char0 & 0xf0) === 0x10) {
+      // unsupported control code
+      } else {
+        // No text will be added to the caption, so just return now
         return;
       }
 
-      // remove null chars
-      if (char0 === 0x00) {
-        char0 = null;
-      }
-      if (char1 === 0x00) {
-        char1 = null;
-      }
-
       // character handling is dependent on the current mode
-      this[this.mode_](packet.pts, char0, char1);
-      break;
+      // note that we only get here when we'll actually be adding text
+      // to the caption
+      this[this.mode_](packet.pts, text);
+
+      break; // ends `default` switch block
```

Comment:
I entirely agree that the `default` case is unwieldy. Honestly, I think the switch should go in favour of if/else blocks. I was hesitant to remove it only because, like you say, it was like that when I got here.

A refactor to if/else would, I think, make a much more logical code flow, because all "handle X" logic would be at the top level. Something like this:

```javascript
var char0 = data >> 8;
var char1 = data & 0x00ff;
if (data == this.RU2) {
  // do roll-up things
} else if (data == this.RCL) {
  // do pop-on things
} else if (char0 == this.specialCharacters && char1 >= 0x30 && char1 <= 0x3f) {
  // append special character
} else if (char0 === this.ext_ && (char1 >= 0x20 && char1 <= 0x2f)) {
  // handle mid-row codes
} else {
  // we've got an unsupported code, so just return
  return;
}
```